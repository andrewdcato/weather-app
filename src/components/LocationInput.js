import React, { useState, useRef } from "react";
import useLoadScript from "../hooks/useLoadScript";

/**
 * Renders place select input and passes lat & lng of
 *  selected place up the tree for forecast lookup
 *
 * @param {*} param0 - funcs for updating lat & lng
 */
export default function LocationInput({ setLat, setLng }) {
  const [query, setQuery] = useState("");
  const autoCompleteRef = useRef(null);

  useLoadScript({
    autoCompleteRef,
    setLat,
    setLng,
    setQuery
  });

  return (
    <>
      <div className="w-full mb-2">
        <div className="flex justify-center">
          <label htmlFor="city" className="sr-only">
            City
          </label>

          <input
            type="text"
            name="city"
            className="w-full block shadow-sm focus:ring-blue-500 focus:border-blue-500 border-gray-300 rounded-md"
            placeholder="Enter a City"
            data-testid='location-input'
            ref={autoCompleteRef}
            onChange={event => setQuery(event.target.value)}
            value={query}
          />
        </div>
      </div>
    </>
  );
}
