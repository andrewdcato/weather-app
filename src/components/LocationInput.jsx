import React, { useState, useEffect, useRef } from "react";
let autoComplete;

/**
 * Dynamically appends script to head element after
 *  component mounts
 *
 * @param {string} url - source of script to load
 * @param {function} callback - func to execute
 */
function loadScript(url, callback) {
  const script = document.createElement("script");
  script.type = "text/javascript";
  script.src = url;

  if (script.readyState) {
    script.onreadystatechange = () => {
      if (['loaded','complete'].includes(script.readyState)) {
        script.onreadystatechange = null;
        callback();
      }
    };
  } else script.onload = () => callback();

  document.getElementsByTagName("head")[0].appendChild(script);
};

/**
 * Handler to hit Places API and update app state w/latitude and
 *  longitude of selected place
 *
 * @param {Object} param0 - funcs for updating app state
 */
function handlePlaceSelect({ setLat, setQuery, setLng }) {
  const place = autoComplete.getPlace();
  const { formatted_address, geometry } = place;

  // Update value of search field
  setQuery(formatted_address);

  // Update lat and lng for weather query
  setLat(geometry.location.lat())
  setLng(geometry.location.lng())
}

/**
 * Sets up autocomplete config and adds listener to place selection
 *  input field
 *
 * @param {Object} param0 - funcs for updating app state
 */
function handleScriptLoad({
  autoCompleteRef, setLat, setLng, setQuery
}) {
  const defaultBounds = new window.google.maps.LatLngBounds();

  autoComplete = new window.google.maps.places.Autocomplete(
    autoCompleteRef.current,
    {
      types: ["(cities)"],
      bounds: defaultBounds,
      componentRestrictions: { country: "us" },
    }
  );

  autoComplete.setFields([
    "address_components", "formatted_address", "geometry"
  ]);

  autoComplete.addListener(
    "place_changed",
    () => handlePlaceSelect({ setLat, setQuery, setLng })
  );
}

/**
 * Renders place select input and passes lat & lng of
 *  selected place up the tree for forecast lookup
 *
 * @param {*} param0 - funcs for updating lat & lng
 */
export default function LocationInput({ setLat, setLng }) {
  const [query, setQuery] = useState("");
  const autoCompleteRef = useRef(null);

  useEffect(() => {
    const { REACT_APP_GOOGLE_API_KEY: API_KEY } = process.env;
    const url = `https://maps.googleapis.com/maps/api/js?key=${API_KEY}&libraries=places`;

    loadScript(
      url,
      () => handleScriptLoad({
        autoCompleteRef, setLat, setLng, setQuery
      }));
  }, [setLat, setLng]);

  return (
    <>
      <input
        type="text"
        placeholder="Enter a City"
        data-testid='location-input'
        ref={autoCompleteRef}
        onChange={event => setQuery(event.target.value)}
        value={query}
      />
    </>
  );
}
