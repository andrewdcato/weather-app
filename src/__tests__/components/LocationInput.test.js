import React from 'react';
import renderer from 'react-test-renderer';
import LocationInput from '../../components/LocationInput';

test('matches snapshot', () => {
  const input = renderer.create(<LocationInput />).toJSON();
  expect(input).toMatchSnapshot();
});
