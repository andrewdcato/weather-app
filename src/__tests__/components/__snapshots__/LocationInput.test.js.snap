// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="w-full mb-2"
>
  <div
    className="flex justify-center"
  >
    <label
      className="sr-only"
      htmlFor="city"
    >
      City
    </label>
    <input
      className="w-full block shadow-sm focus:ring-blue-500 focus:border-blue-500 border-gray-300 rounded-md"
      data-testid="location-input"
      name="city"
      onChange={[Function]}
      placeholder="Enter a City"
      type="text"
      value=""
    />
  </div>
</div>
`;
